### PR TITLE
machines: Extend for qemu:///session

### PR DESCRIPTION
**Pull request by mareklibra**
Libvirt's connection qemu:///session is added to existing qemu:///system
as a data source for the package.

The data from both sources are merged to a single list.

Due to libvirt limitation, the 'root' user has access
to qemu:///system only.

### CODE REVIEW

**Comment by mareklibra** [comment]
qemusession

**Comment by mareklibra** [comment]
The user receives a single list with both sources (means connections) merged.
Internally, both qemu:///system and qemu:///session are retrieved independently.

The user shall be able to differentiate between them, especially because connections do not share single namespace.
So far, the 'connection' is listed as a 'high level' property along the name and state, visible even with the row collapsed.

I was thinking about either colors (for 'name' or even the whole row) or icons close to the VM name.
On the other, the list might be sorted by connection on day, once list sorting is implemented.

Or render both sources to two separate lists? I like the idea of single list more.

@garrett, @andreasn, what do you think, please?

**Comment by andreasn** [comment]
What is the difference between user and system for the end user?
What are the different cases where you need to differentiate between them?

**Comment by mareklibra** [comment]
The qemu:///system conforms to a _single_ process running on the system under root permissions.
It has full access to resources, granting the required ones to the VM (means qemu process) which is actually owned by the 'qemu' user (so usually without access to user's files, etc).

The qemu:///session conforms to a libvirtd process owned by the user, not easily accessible by other users (including the root). Usually used on desktops due to sharing permissions with the logged user starting the VM.

For the end user: 
- session's VM might be limited in privileges for resources and is not visible (more or less) for other users. Managed by the user itself.
- system's VM might not be related to the particular user, might be started by more complex virt manager like oVirt, it's life does not correspond to the user session and the user might not have full management privileges for it (like Shutdown). The user might still need to get connection to it (like console), check address or monitor resources. Or just know about it's existence before system shutdown.

Single list gives better general overview 'What's happening on the system' but for particular actions the user needs to understand context (user- or system-wide).

File: pkg/machines/config.es6 [hunk 1]
```diff
@@ -25,7 +25,14 @@
 const VMS_CONFIG = {
     DefaultRefreshInterval: 10000, // in ms
     Virsh: {
-        ConnectionParams: ['-c', 'qemu:///system']
+        connections: {
+            'system': {
```
Comment (stefwalter) [inline pkg/machines/config.es6]: Are these translatable connection names? Are they translated in virt-manager or elsewhere?

Comment (stefwalter) [inline pkg/machines/config.es6]: If marked as translatable, these will need to use double quotes.

File: pkg/machines/libvirt.es6 [hunk 2]
```diff
@@ -99,59 +99,82 @@ export default {
...
 
+function canLoggedUserConnectSession (connectionName, loggedUser) {
+    return connectionName !== 'session' || loggedUser.name !== 'root';
+}
+
+function doGetAllVms (dispatch, connectionName) {
+    const connection = VMS_CONFIG.Virsh.connections[connectionName];
+
+    return spawnScript({
+        script: `virsh ${connection.params.join(' ')} -r list --all | awk '$1 == "-" || $1+0 > 0 { print $2 }'`
```
Comment (stefwalter) [inline pkg/machines/libvirt.es6]: Does this account for escaping? Does it need to?

File: pkg/machines/services.es6 [hunk 3]
```diff
@@ -24,7 +24,7 @@ export function spawnProcess({ cmd, args = [], stdin}) {
     const spawnArgs = [cmd, ...args];
     logDebug(`spawn process args: ${spawnArgs}`);
 
-    return spawn(cockpit.spawn(spawnArgs, { superuser: "try", err: "message" })
+    return spawn(cockpit.spawn(spawnArgs, { /* superuser: "try",*/ err: "message" })
```
Comment (stefwalter) [inline pkg/machines/services.es6]: Lets remove, rather than comment this out.

File: pkg/machines/libvirt.es6 (hunk 2 above)
Comment (mareklibra) [inline pkg/machines/libvirt.es6]: Only values from config.es6 can appear here, so I don't think escaping is needed.

File: pkg/machines/services.es6 (hunk 3 above)
Comment (mareklibra) [inline pkg/machines/services.es6]: Done

File: pkg/machines/config.es6 (hunk 1 above)
Comment (mareklibra) [inline pkg/machines/config.es6]: Adding translation to helper.es6:rephraseUI() since these connection-name-strings are visible for the user.

**Comment by garrett** [comment]
Hi! I think I need to better understand the thought behind this feature before recommending what to do with the merged list of machines (whether it should stay as-is, be filtered, or split into two lists). After looking at the mockup and what's written about it in this issue, I have a few questions:

1. Most people be using system _or_ user VMs — and not both. Is this correct?

2. How common will it be for someone to have _both_ system and user VMs running on the same machine?

3. In the cases a person might use both types on the same machine, why would they do so? (Especially considering the machine in question would most likely be a server, not a workstation, as they're using Cockpit.)

4. I don't see a way to interact with the VMs currently. Is that coming, or does one have to click and expand each row to perform actions like reboot, shut down, & delete?

**Comment by mareklibra** [comment]
TL;DR: yes, it's not uncommon to have both types running on a single machine.

The system VMs are expected more on a server, but not exclusively. System VMs have easier access to resources of the hosting machine. High-level management apps, like oVirt, start this type of VMs only. It's common for standard or wheel user to start system VMs, i.e. via virt-manager.

The 'session VMs' are started either if the user lacks access to the qemu:///system or if the VM (means QEMU) needs to share permissions with the logged user. An example for that are gnome-boxes. But it's common to start session VM via virt-manager, it's just up to the user decision.

For completeness, the root does not need to have automatically full access to the qemu:///system. He usually has, but it might be configured otherwise. Like for example in the oVirt environment.

On non-single-purpose servers, I would expect both types running - multiple user accounts exist or because of the app isolation.

**Comment by mareklibra** [comment]
Regarding the interaction with a VM: the actions are accessible after row expansion. So far there are just 5 actions but the list will grow in the future. Maybe the subtabs will hold specific actions.

There was an attempt to make the shutdown-, reset-actions visible even in collapsed state but it was not visually be looking good.

**Comment by garrett** [comment]
I apologize for this deluge of questions; I want to make sure I understand how VMs in Cockpit are handled. I do understand the technical difference between the types, but am still not quite sure about the circumstances where someone would use both types on a server. (I'm sure it happens; I would just like to know _why_ someone would mix types on the same server.)

- Would someone with both session and system VMs want to administer both in the same way?
    - At least for typical operations, like start/stop/reboot/remove, I'd guess the answer is yes.

- If it's a session VM… then knowing who launched it might be important too, right?
    - It could matter for system VM too — but that would probably be more difficult to ascertain, as it would most likely shown to be oVirt or some other VM manager.
    - I'm guessing many people using the system might launch the same VM as a starting point and, aside from who launched it, it might look nearly identical inside Cockpit's current list view of VMs.

- How many VMs would be shown at the same time within Cockpit?
  - Should the list be filterable across multiple items of metadata (user, VM name, VM file name, etc.)?

- Might Cockpit have the ability to launch VMs in the future?
    - And would it be system and/or session VMs?

- What's the ideal scope of Cockpit's handling of qemu-based VMs?
    - When would someone be able to rely on Cockpit _only_ versus having to graduate to something like oVirt to do what they need?
    - _Note: Not talking about current implementation here, but an ideal implementation of VMs within Cockpit._

---

BTW, about necessary privileges for VMs and migration:

At least from my understanding, it seems like someone would be fine with session VMs unless they need USB/PCI device assignments, more extensive virtual networks, and more types of remote access via libvirtd — and, of course, they would require special privileges to get these features.

It does seem like [migration from session to system VMs is possible](https://mike42.me/blog/libvirt-migrate-a-vm-from-qemusession-to-qemusystem), but permissions on the files would need to be changed and the VM-related should probably be moved to wherever the system VMs are stored.

**Comment by mareklibra** [comment]
| - Would someone with both session and system VMs want to administer both in the same way?
  Yes, I think so.

| If it's a session VM… then knowing who launched it might be important too, right?
  Very good point. Recently, session VMs of logged user are listed only.
  There's no easy way to get list of all session VMs on the host since libvirt does not support that. There are multiple workarounds, but nothing non-hacking.

  If this is going to be implemented, I would suggest to differentiate on the 'connection-name' level, like 'Session-[MyUserName]'.

  So, recently all listed VMs are either owned by the user or come from qemu:///system.

| It could matter for system VM too — but that would probably be more difficult to ascertain, as it would most likely shown to be oVirt or some other VM manager.
  There's no reliable way to get info about the originator. But for selected cases, that might be a very nice feature.

| I'm guessing many people using the system might launch the same VM as a starting point and, aside from who launched it, it might look nearly identical inside Cockpit's current list view of VMs.
  As mentioned above, the session VMs of different users will not be mixed. In case of the system VMs: the name is unique.

| How many VMs would be shown at the same time within Cockpit?
  Good point. In extreme, even hundreds. On x86 servers, the count is in tens.
  Of course, there might be may VMs defined but not running - their number is (almost) unlimited.
  
  Recently there's no paging nor filtering. I have this feature on mind, but it's not yet implemented. If you can design it, it will be very helpful.


| Might Cockpit have the ability to launch VMs in the future?
  Launching a VM is two step process: define and start.
  The 'start' is already there.
  The 'define' is tricky one. I would love to have it there but it's not a high priority now due to complexity and existence of other tools for workaround.
  Topic for discussion: shall it be complex like the virt-manager or as simple as gnome-boxes from user perspective?

  Anyway: yes, let's assume the VM creation will be there in the future. And once it's done, both system and session make sense, I would keep that decision up to the user and privileges.


| When would someone be able to rely on Cockpit only versus having to graduate to something like oVirt to do what they need?

  The oVirt is an ideal solution if the user need to manage, monitor, migrate higher number of VMs over multiple hosts. In addition, the oVirt handles related complex settings like storage or networks. So oVirt has benefits in an environment of two or more hosts running the VMs.

  But even with oVirt, the Cockpit has benefits in fine-tuning and monitoring while providing host-centric view. This is hard to achieve in high-level management tool, but Cockpit runnig on the particular host is ideal.

  It's still an open question, but what if the VM management in Cockpit and (especially) other infrastructure grows to the point where the oVirt cluster is distributively managed from Cockpit running on a cluster-node?

**Comment by stefwalter** [comment]
@garrett Any objections to merging this, and doing further work as a follow up?

**Comment by garrett** [comment]
@stefwalter: I think that's a great suggestion. This looks good to me for merging as-is, and would like to see the additional points we all brought up here addressed in the future.

**Comment by mareklibra** [comment]
@dperpeet, done

**Comment by dperpeet** [comment]
Looks good, ready to merge once tests pass.

**Comment by dperpeet** [comment]
remaining test failures are unrelated
